File: libraries/cyclestreets-core/src/main/java/net/cyclestreets/AppInfo.java
package net.cyclestreets;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

public class AppInfo {
  private static String version_ = null;

  public static String version(final Context context) {
    if (version_ == null)
      version_ = String.format("%s/%s", context.getPackageName(), versionName(context));
    return version_;
  }

  private static String versionName(final Context context) {
    try {
      final PackageInfo info = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
      return String.format("%s/%s", info.versionName, String.valueOf(info.versionCode));
    }
    catch (PackageManager.NameNotFoundException nnfe) {
      // like this is going to happen
      return UNKNOWN;
    }
  }

  private static final String UNKNOWN = "unknown";
}

File: libraries/cyclestreets-core/src/main/java/net/cyclestreets/AppInfo.kt
package net.cyclestreets

import android.content.Context
import android.content.pm.PackageManager

private const val UNKNOWN = "unknown"

object AppInfo {
    private var version: String? = null

    fun version(context: Context): String {
        if (version == null)
            version = String.format("%s/%s", context.packageName, versionNameAndCode(context))
        return version!!
    }

    private fun versionNameAndCode(context: Context): String {
        return try {
            val info = context.packageManager.getPackageInfo(context.packageName, 0)
            String.format("%s/%s", info.versionName, info.versionCode.toString())
        } catch (e: PackageManager.NameNotFoundException) {
            // like this is going to happen
            UNKNOWN
        }
    }
}
